### DISCUSSION TITLE
react on `ObserveCollectionChanges()` completion

### DISCUSSION THREAD

**Discussion by nor0x** [reactivemarbles/DynamicData#921] [Q&A]
I'm getting started wiht DynamicData and In my code i have a collection of items `ReadonlyObservableCollection<T> items;`

and subscribe to changes to that collection to react on the specific changes like this

```csharp
items.ObserveCollectionChanges()
.Subscribe(e =>
{
	//do stuff based on e.EventArgs.Action
});
```

I wanted to know if there is a way to get notified once the CollectionChanges are completed? Is this something that is built-in to DynamicData?

Thank you!

**Comment by JakenVeina** [comment]
Folks will usually tell you to avoid side-effects in RX streams, and it's good advice, but that's essentially what you want here. If you have multiple steps of work to complete, and you can't just do all the work within a single direct subscription, you can either build your pre-requisite work int your observable stream, as side-effects, or you can initiate an entirely new downstream.

```cs
var consumableStream = items
    .ObserveCollectionChanges()
    .Do(e =>
    {
        //do stuff based on e.EventArgs.Action
    });
    
// You can now hand-off consumableStream to be subscribed to by whoever else needs to do work, after your processing in .Do() is done.
```

```cs
using var changesProcessed = new Subject<Unit>();

using var itemsSubscription = items
    .ObserveCollectionChanges()
    .Subscribe(e =>
    {
        //do stuff based on e.EventArgs.Action
        
        changesProcessed.OnNext(Unit.Default);
    });

// You can now hand-off changesProcessed to be subscribed to by whoever else needs to do work, after your processing in .Subscribe() is done.
```

**Reply by Snailya** [reply]
What are your thoughts on modifying the observable item itself ​​in side-effects? For example, the item has a navigation property that is not load yet, then calling the service to fill that proeprty in side-effects.

**Reply by JakenVeina** [reply]
I would say it's not fundamentally different than any other type of side-effect. Generally best avoided, but if you're going to do it, be very cognizant of how that might cause things to happen back upstream. Cause, that's really the real danger with side-effects, having them inadvertently trigger additional upstream actions, creating a loop. That tends to result in either infinite looping, or deadlocks.
